File: primes/primes.rb
def n_primes(n)
  # returns an array of the first n primes
end

def primes_up_to(n)
  # returns an array of primes up to (and including) n
  return [] if n < 2
  @primes = [2]
  return @primes if n == 2
  next_prime = 3
  while @primes.last < n
    @primes << next_prime
    next_prime = find_next_prime(next_prime + 2)
  end
  @primes
end

def find_next_prime(x)
  return x if is_prime?(x)
  find_next_prime(x + 2)
end

def is_prime?(x)
  @primes.each do |prime|
    return false if x % prime == 0
  end
  return true
end


puts primes_up_to(0) == []
puts primes_up_to(1) == []
puts primes_up_to(2) == [2]
puts primes_up_to(7) == [2,3,5,7]
puts primes_up_to(47) == [2,3,5,7,11,13,17,19,23,29,31,37,41,43,47]

File: dbc_talks/primes/primes.js
var nPrimes = function(n) {
  // returns a list of the first n primes
  if (n === 0) { return []; }
  if (n === 1) { return [2]; }
  var primes = [2];
  nextPrime = 3;
  while (primes.length < n) {
    primes.push(nextPrime);
    nextPrime = findNextPrime(primes);
  }
  return primes;
}

var primesUpTo = function(n) {
  // returns a list of primes up to n
  // This is implemented in Ruby filem so left blank here
}

var findNextPrime = function(primes) {
  x = primes.last() + 2;
  while (!isPrime(primes, x)) {
    x += 2;
  }
  return x;
}

var isPrime = function(knownPrimes, x) {
  for (i=0; i<knownPrimes.length; i++) {
    if (x%knownPrimes[i] === 0) {
      return false;
    }
  }
  return true;
}

Array.prototype.last = function() {
  return this[this.length - 1];
}
Array.prototype.eq = function(other) {
  if (this.length != other.length) {
    return false;
  }
  for (i=0; i<this.length; i++) {
    if (this[i] != other[i]) {
      return false;
    }
  }
  return true;
}

console.log(nPrimes(2).eq([2,3]));
console.log(nPrimes(1).eq([2]));
console.log(nPrimes(0).eq([]));
console.log(nPrimes(10).eq([2,3,5,7,11,13,17,19,23,29]))
